### PR TITLE
feat: Change the sessions matview to use pre-aggregated counts (3/4)

### PR DESCRIPTION
**Pull request by Swatinem**

### CODE REVIEW

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    release,
+    environment,
+    user_agent,
+    os,
+    runtime,
+
+    -- we would have liked to changed the quantiles here to `0.5, 0.75, 0.95, 0.99`:
+    quantilesIfState(0.5, 0.9)(
+        duration,
+        duration <> {MAX_UINT32} AND status == 1
```
Comment (fpacifici) [inline snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:66]: Shouldn't you exclude the preaggregated sessions ? Or is it implicitly done by `status == 1`? If the condition is implicit please add a comment about it.

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    ) as duration_quantiles,
+
+    -- this is new:
+    avgIfState(duration, duration <> {MAX_UINT32} AND status == 1) as duration_avg,
+
+    -- use the old `count` columns for "legacy" rows with `quantity == 0`:
+    countIfState(session_id, seq == 0 AND quantity == 0) as sessions,
+    countIfState(session_id, status == 2 AND quantity == 0) as sessions_crashed,
+    countIfState(session_id, status == 3 AND quantity == 0) as sessions_abnormal,
```
Comment (fpacifici) [inline snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:75]: As mentioned in the first PR in this stack. I think we can just remove these and avoit the artificial condition on `quantity == 0`. Clickhouse will not fill these columns and they will be treated as empty.

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                table_name="sessions_hourly_mv_local",
+            ),
+            create_matview_v1,
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.Operation]:
+        return []
+
+    def backwards_dist(self) -> Sequence[operations.Operation]:
+        return []
```
Comment (fpacifici) [inline snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:139]: missing

Comment (Swatinem) [reply snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:130]: Are these needed for the matview itself? The initial `0001` migration does not have these either.

File: snuba/migrations/snuba_migrations/sessions/0001_sessions.py [hunk 4]
```diff
@@ -11,7 +10,8 @@
 from snuba.clusters.storage_sets import StorageSetKey
 from snuba.migrations import migration, operations, table_engines
 from snuba.migrations.columns import MigrationModifiers as Modifiers
-from snuba.processor import MAX_UINT32, NIL_UUID
+from .matview import aggregate_columns_v1, create_matview_v1
```
Comment (Swatinem) [inline snuba/migrations/snuba_migrations/sessions/0001_sessions.py:13]: you were commenting on this in one of the previous PRs. I moved the code to a common module since I want to dedupe it with the `0003` backward statement.
If this is not desired, I can revert this and rather copy-paste the first matview into the rollback statement.

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py (hunk 3 above)
Comment (fpacifici) [reply snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:131]: ok, ignore then. I thought we had the DDL for the matview in the distributed migration.

**Review by fpacifici — COMMENTED** [review]
This step will be irreversible once applied so we need a better test plan.
1) Please ensure the test on sentry side for sessions all pass when hitting snuba with this PR applied. And note here how you ran those tests.
2) I would advise we add a snuba integration test for sessions that is on the line of test_api.py and test_discover_api.py. This test would actually store some rows in clickhosue and hit the snuba api endpoint to ensure the query returns the proper results. Since you do not have a consumer for preaggregated sessions yet, you may want to manually add those lines in clickhouse from the test. This is an example of how to manually write rows in clickhouse from a test: https://github.com/getsentry/snuba/pull/1525/files#diff-7cc40f1cb85a8f19c80b2e901626d78e4729047dbbb0ed52861f5894ff688115R158-R161
3) We need some manual test performed before merge and that James will be able to quickly perform after the migration and before restarting the consumer. Could you provide an INSERT statement for non aggregated sessions and one for aggregated sessions and a SELECT statement as it would be generated by the product to get the combined results ?

**Comment by Swatinem** [comment]
Been running the things noted here:
https://github.com/getsentry/snuba#testing-against-sentry

* `USE_SNUBA=1 make test-snuba`
* `USE_SNUBA=1 make test-acceptance`
  this had 2 failures: `OrganizationEventsV2Test.test_event_detail_view_from_transactions_query` in `tests/acceptance/test_organization_events_v2.py` and `PerformanceSummaryTest.test_view_details_from_summary` in `tests/acceptance/test_performance_summary.py`. I will re-run these in a minute since I saw some intermittent problems with this testsuite before as well.
* `make test-python`

**Comment by Swatinem** [comment]
Re-running once, `OrganizationEventsV2Test.test_event_detail_view_from_transactions_query` was still failing. Re-running just that one file and it succeeds, so that seems to be very flacky.

**Comment by Swatinem** [comment]
I think these queries can be used as tests:
```sql
INSERT INTO	sessions_raw_local
	(session_id, distinct_id, quantity, seq, duration, status, errors, started)
VALUES
	('00000000-0000-0000-0000-000000000123', '00000000-0000-0000-0000-000000000000', 1, 0, 4294967295, 0, 0, '2020-12-16 13:25:37'),
	('00000000-0000-0000-0000-000000000123', '00000000-0000-0000-0000-000000000000', 1, 123, 4294967295, 1, 1, '2020-12-16 13:25:37'),
	('00000000-0000-0000-0000-000000000000', '00000000-0000-0000-0000-000000000000', 4, 0, 4294967295, 4, 0, '2020-12-16 13:25:37'),
	('00000000-0000-0000-0000-000000000000', '00000000-0000-0000-0000-000000000123', 5, 0, 4294967295, 1, 0, '2020-12-16 13:25:37');
	
SELECT
	(plus(countIfMerge(sessions), sumIfMerge(sessions_preaggr)) AS _snuba_sessions),
	(plus(uniqIfMerge(sessions_errored), sumIfMerge(sessions_errored_preaggr)) AS _snuba_sessions_errored),
	(uniqIfMerge(users) AS _snuba_users),
	(uniqIfMerge(users_errored) AS _snuba_users_errored)
FROM sessions_hourly_local
PREWHERE in(project_id, tuple(0))
WHERE
	equals(org_id, 0) AND greaterOrEquals(started, '2020-12-16 13:00:00') AND less(started, '2020-12-16 14:00:00')
LIMIT 1000 OFFSET 0;
```

The expected result is:
```
_snuba_sessions|_snuba_sessions_errored|_snuba_users|_snuba_users_errored|
---------------|-----------------------|------------|--------------------|
             10|                      5|           1|                   0|
```

File: tests/test_sessions_api.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            dict(template, quantity=9, status=4),
+            # pre-aggregated "exited" sessions with user
+            dict(template, quantity=5, distinct_id=user_2, status=1),
+            # pre-aggregated "exited" session
+            dict(template, quantity=4, status=1),
+        ]
+
+        cluster.get_batch_writer(
+            "sessions_raw_local", DummyMetricsBackend(strict=True), None, None
+        ).write([json.dumps(session).encode("utf-8") for session in sessions])
```
Comment (fpacifici) [inline tests/test_sessions_api.py:85]: Please load a storage and get the batch writer from there. This will abstract the table name from there.
```
from snuba.datasets.storages.factory import get_writable_storage

storage = get_writable_storage(StorageKey::SESSIONS_RAW)
storage.get_table_writer().get_batch_writer().write(...)
```

File: tests/test_sessions_api.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        self.generate_manual_session_events()
+
+    def teardown_method(self, test_method):
+        # Reset rate limits
+        state.delete_config("global_concurrent_limit")
+        state.delete_config("global_per_second_limit")
+        state.delete_config("project_concurrent_limit")
+        state.delete_config("project_concurrent_limit_1")
+        state.delete_config("project_per_second_limit")
+        state.delete_config("date_align_seconds")
```
Comment (fpacifici) [inline tests/test_sessions_api.py:39]: Why are these needed ?

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    -- pre-aggregated sessions dont have a duration, so no more filtering is needed.
+    -- we would have liked to changed the quantiles here to `0.5, 0.75, 0.95, 0.99`:
+    quantilesIfState(0.5, 0.9)(
+        duration,
+        duration <> {MAX_UINT32} AND status == 1
+    ) as duration_quantiles,
+
+    -- this is new:
+    avgIfState(duration, duration <> {MAX_UINT32} AND status == 1) as duration_avg,
```
Comment (fpacifici) [inline snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:69]: Does that imply that pre-aggregated session never have a duration. If that's the case please explain.

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    -- `sum` the session counts based on the new `quantity`:
+    sumIfState(quantity, seq == 0) as sessions_preaggr,
+    sumIfState(quantity, status == 2) as sessions_crashed_preaggr,
+    sumIfState(quantity, status == 3) as sessions_abnormal_preaggr,
+
+    -- individual session updates keep using the uniq session_id as before:
+    uniqIfState(session_id, errors > 0 AND session_id != '{NIL_UUID}') as sessions_errored,
+
+    -- pre-aggregated counts use sum. by definition, `crashed` and `abnormal` are errored:
+    sumIfState(quantity, status IN (2, 3, 4) AND session_id == '{NIL_UUID}') as sessions_errored_preaggr,
```
Comment (fpacifici) [inline snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:80]: Is there any session coming in with session_id == NIL_UUID as of now that would be counted in this counter ?

**Review by fpacifici — COMMENTED** [review]
Seems mostly good to me, will prepare the DDL for James. Could you have a look at my questions before ?

**Comment by fpacifici** [comment]
@JTCunning this is the DDL for the new materialized view. There are still some questions on the PR before we consider running this but none of those would impact the ddl:

```
 CREATE MATERIALIZED VIEW default.sessions_hourly_mv_local 
 TO default.sessions_hourly_local (
    `org_id` UInt64, 
    `project_id` UInt64, 
    `started` DateTime, 
    `release` LowCardinality(String), 
    `environment` LowCardinality(String), 
    `user_agent` LowCardinality(String), 
    `os` LowCardinality(String), 
    `duration_quantiles` AggregateFunction(quantilesIf(0.5, 0.9), UInt32, UInt8), 
    `duration_avg` AggregateFunction(avgIf, UInt32, UInt8), 
    `sessions` AggregateFunction(countIf, UUID, UInt8), 
    `sessions_preaggr` AggregateFunction(sumIf, UInt32, UInt8), 
    `sessions_crashed` AggregateFunction(countIf, UUID, UInt8), 
    `sessions_crashed_preaggr` AggregateFunction(sumIf, UInt32, UInt8), 
    `sessions_abnormal` AggregateFunction(countIf, UUID, UInt8), 
    `sessions_abnormal_preaggr` AggregateFunction(sumIf, UInt32, UInt8), 
    `sessions_errored` AggregateFunction(uniqIf, UUID, UInt8), 
    `sessions_errored_preaggr` AggregateFunction(sumIf, UInt32, UInt8), 
    `users` AggregateFunction(uniqIf, UUID, UInt8), 
    `users_crashed` AggregateFunction(uniqIf, UUID, UInt8), 
    `users_abnormal` AggregateFunction(uniqIf, UUID, UInt8), 
    `users_errored` AggregateFunction(uniqIf, UUID, UInt8)
) 
AS SELECT 
    org_id, 
    project_id, 
    toStartOfHour(started) AS started, 
    release, environment, 
    user_agent, 
    os,
    quantilesIfState(0.5, 0.9)(duration, (duration != 4294967295) AND (status = 1)) AS duration_quantiles, 
    avgIfState(duration, (duration != 4294967295) AND (status = 1)) AS duration_avg, 
    sumIfState(quantity, seq = 0) AS sessions_preaggr, 
    sumIfState(quantity, status = 2) AS sessions_crashed_preaggr, 
    sumIfState(quantity, status = 3) AS sessions_abnormal_preaggr, 
    uniqIfState(session_id, (errors > 0) AND (session_id != '00000000-0000-0000-0000-000000000000')) AS sessions_errored, 
    sumIfState(quantity, (status IN (2, 3, 4)) AND (session_id = '00000000-0000-0000-0000-000000000000')) AS sessions_errored_preaggr, 
    uniqIfState(distinct_id, distinct_id != '00000000-0000-0000-0000-000000000000') AS users, 
    uniqIfState(distinct_id, (status = 2) AND (distinct_id != '00000000-0000-0000-0000-000000000000')) AS users_crashed, 
    uniqIfState(distinct_id, (status = 3) AND (distinct_id != '00000000-0000-0000-0000-000000000000')) AS users_abnormal, 
    uniqIfState(distinct_id, (errors > 0) AND (distinct_id != '00000000-0000-0000-0000-000000000000')) AS users_errored 
FROM default.sessions_raw_local 
GROUP BY org_id, project_id, started, release, environment, user_agent, os
```

(taken from clickhouse after running the migraiton locally)


Now this is a risky migration as it would be basically impossible to revert after we restart the consumer because we would start filling the `sum` columns and stop filling the old ones, so reverting without loosing data would be basically impossible unless we rewinded the kafka offset, but still we the new data would be aggregated and we could not remove it from the table.

I think we should go this way:
1) stop consumer
2) apply the DDL the way you prefer
3) run the manual test Arpad provided above (https://github.com/getsentry/snuba/pull/1487#issuecomment-746299484) to ensure everything works properly 
4) only at that point restart the consumer.

Do you prefer planning this before the lockdown or after? I think it should be done in a moment when both me and arpad are available (I am flexibly with timing)

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py (hunk 8 above)
Comment (Swatinem) [reply snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:80]: Relay will enforce a uuid for individual session updates.

File: snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py (hunk 7 above)
Comment (Swatinem) [reply snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py:69]: Its stated for the quantiles above, but I copied this down as well

File: tests/test_sessions_api.py (hunk 6 above)
Comment (Swatinem) [reply tests/test_sessions_api.py]: I just copy-pasted those from `test_api.py`/`test_transactions_api.py`, I removed them and tests seem to be happy.

**Comment by fpacifici** [comment]
This depends on https://github.com/getsentry/snuba/pull/1579 to be merged before it can go ahead.

**Comment by fpacifici** [comment]
Some thorough test queries. 
Both queries compare the aggregations computed from the raw table with the same aggregation taken from the hourly table.
In order to do that they join the time bucket in both tables and compare the results.



```
# For each column, it gets the raw data, the hourly (old column) and the hourly (new column), the raw
# data must correspond to the sum of the old and new column (well kind of).
SELECT 
    raw.started, 
    'TOTAL',
    raw.sessions as raw_total, 
    hourly.sessions as hourly_old,
    hourly.sessions_preaggr as hourly_new,
    (hourly.sessions + hourly.sessions_preaggr) as hourly_total,    
    (hourly_total/raw_total) as total_error,
    'CRASHED',
    raw.sessions_crashed as raw_crashed, 
    hourly.sessions_crashed as hourly_crashed_old,
    hourly.sessions_crashed_preaggr as hourly_crashed_new,
    (hourly.sessions_crashed + hourly.sessions_crashed_preaggr) as hourly_crashed,
    (hourly_crashed/raw_crashed) as crashed_error,
    'ABNORMAL',
    raw.sessions_abnormal as raw_abnormal, 
    hourly.sessions_abnormal as hourly_abnormal_old,
    hourly.sessions_abnormal_preaggr as hourly_abnormal_new,
    (hourly.sessions_abnormal + hourly.sessions_abnormal_preaggr) as hourly_abnormal,
    (hourly_abnormal/raw_abnormal) as abnormal_error,
    'ERRORED',
    raw.sessions_errored as raw_errored, 
    hourly.sessions_errored as hourly_errored_old,
    hourly.sessions_errored_preaggr as hourly_errored_new,
    (hourly.sessions_errored + hourly.sessions_errored_preaggr) as hourly_errored,
    (hourly_errored/raw_errored) as errored_error
FROM (
    SELECT 
        toStartOfHour(started) AS started,
        countIf(session_id, seq == 0) as sessions,
        countIf(session_id, status == 2) as sessions_crashed,
        countIf(session_id, status == 3) as sessions_abnormal,
        uniqIf(session_id, errors > 0) as sessions_errored
    FROM sessions_raw_dist
    PREWHERE started = toStartOfHour(now())
    GROUP BY started
) raw 
INNER JOIN (
    SELECT 
        started,
        countIfMerge(sessions) as sessions,
        sumIfMerge(sessions_preaggr) as sessions_preaggr,
        countIfMerge(sessions_crashed) as sessions_crashed,
        sumIfMerge(sessions_crashed_preaggr) as sessions_crashed_preaggr,
        countIfMerge(sessions_abnormal) as sessions_abnormal,
        sumIfMerge(sessions_abnormal_preaggr) as sessions_abnormal_preaggr,
        uniqIfMerge(sessions_errored) as sessions_errored,
        sumIfMerge(sessions_errored_preaggr) as sessions_errored_preaggr
    FROM sessions_hourly_dist
    PREWHERE started = toStartOfHour(now())
    GROUP BY started
) hourly
ON raw.started = hourly.started
FORMAT Vertical


SELECT 
    raw.started, 
    'TOTAL',
    raw.users as raw_users,
    hourly.users as hourly_users,
    'CRASHED',
    raw.users_crashed as raw_users_crashed,
    hourly.users_crashed as hourly_users_crashed,
    'ABNORMAL',
    raw.users_abnormal as raw_users_abnormal,
    hourly.users_abnormal as hourly_users_abnormal,
    'ERRORED',
    raw.users_errored as raw_users_errored,
    hourly.users_errored as hourly_users_errored
FROM (
    SELECT 
        toStartOfHour(started) AS started,
        uniqIf(distinct_id, distinct_id != '00000000-0000-0000-0000-000000000000') as users,
        uniqIf(distinct_id, status == 2 AND distinct_id != '00000000-0000-0000-0000-000000000000') as users_crashed,
        uniqIf(distinct_id, status == 3 AND distinct_id != '00000000-0000-0000-0000-000000000000') as users_abnormal,
        uniqIf(distinct_id, errors > 0 AND distinct_id != '00000000-0000-0000-0000-000000000000') as users_errored
    FROM sessions_raw_dist
    PREWHERE started = toStartOfHour(now())
    GROUP BY started
) raw 
INNER JOIN (
    SELECT 
        started,
        uniqIfMerge(users) as users,
        uniqIfMerge(users_crashed) as users_crashed,
        uniqIfMerge(users_abnormal) as users_abnormal,
        uniqIfMerge(users_errored) as users_errored
    FROM sessions_hourly_dist
    PREWHERE started = toStartOfHour(now())
    GROUP BY started
) hourly
ON raw.started = hourly.started
FORMAT Vertical
```

**Review by fpacifici — APPROVED** [review]

**Comment by fpacifici** [comment]
We successfully applied the migration.
